### ISSUE TITLE
Add a self-update mechanism

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-upgrader#130]
## Acceptance Criteria

 * As a developer on an outdated version of a phar or composer managed upgrader, I can find out that I need to update when I next run the command
 * As a developer on an outdated version, I can upgrade the phar by running a command on my local machine
 * There's some protections against upgrading to malicious code
 * The upgrade process can safely roll back in case of errors
 * As an upgrader module maintainer, I can create a new release without error prone manual steps around bumping the version number
 * The release process is well documented for maintainers

## Notes

 * Excludes self updating installations through composer (rather than phar)
 * Excludes code signing
 * Should not require new infrastructure to host phar files
 * Phar files should be served via a secure TLS connection
 * Should not require additional configuration by developers (e.g. Github API keys)

## Original Issue

It would be cool if we had a mechanism to self-update the upgrader. This would involve:
* publicising the latest version of the upgrader ;
* adding logic to the upgrader that allows it to detect the latest version and notify the user if a new release is available ;
* adding a `self-update` command that downloads the latest version.

We need a bit of split logic between the PHAR and composer version, as the update path between those two distribution channel would differ.

It could be useful if you can provide a specific version number as an argument.

## Extra tasks
* [x] Update upgrade guide about how to upgrade the upgrader
* [x] Ask comms to tweet about new updated upgrader

# Pull Requests
* [x] #153 
* [ ] https://github.com/silverstripe/silverstripe-framework/pull/8796

**Comment by dhensby** [comment]
To add to this:

- modify the release builder on travis to publish a file hash of the phar to allow remote verification of its integrity

**Comment by maxime-rainville** [comment]
This is kind of dependent on #148

**Comment by maxime-rainville** [comment]
There's a few project out there meant to make this easy:
* https://github.com/humbug/phar-updater
* https://github.com/consolidation/self-update

**Comment by maxime-rainville** [comment]
#153 is ready for review. Here's how the various ACs are addessed.

> As a developer on an outdated version of a phar or composer managed upgrader, I can find out that I need to update when I next run the command

I've added a bit of code that will check what the latest release is and invite you to update if necessary. This will run on any command, but will only run once. There's a flag save in memory. When you reboot your computer, that flag will be gone and you'll get the message again.

> As a developer on an outdated version, I can upgrade the phar by running a command on my local machine

There's a bit of split logic. If you are running a PHAR, you can update directly via self-update. Otherwise, you'll get a warning inviting you to run a `composer global update`.

> There's some protections against upgrading to malicious code

The phar will be downloaded over HTTPS from GitHub. I had a look at signing the actual PHAR file, but that include shipping a public key along with your PHAR and we would need to create a specialised installation script a bit like the one composer uses. So it's probably another mission for another day.

> The upgrade process can safely roll back in case of errors

The library that I'm using for this comes with a rollback feature. I've just use this.

> As an upgrader module maintainer, I can create a new release without error prone manual steps around bumping the version number

The travis build will embed a .env file into with a version number based on the tag when creating a new release. 

We still need to update the version manually for people installing the upgrader with composer. I didn't find any elegant way of reading this dynamically.

> The release process is well documented for maintainers
I've updated `doc/en/releasing.md`.

# Other things
* I've updated the build script to preserve the older PHAR builds and compute MD5 hashes.

**Comment by maxime-rainville** [comment]
Main PR is merged. I've added a blur to https://github.com/silverstripe/silverstripe-upgrader/pull/156 to encourage people to update their upgrader install.

**Comment by maxime-rainville** [comment]
I've created a separate task to create a blog post and a blurb as been added to the doc in a related PR. I'm calling this done.
